VariantShredding.md

```diff
@@ -25,290 +25,316 @@
-It would need to be written to the `variant_value` column.
+For example, a Variant `event` field may shred `event_type` (`string`) and `event_ts` (`timestamp`) columns using the following definition:
+```
+optional group event (VARIANT) {
+  required binary metadata;
+  optional binary value;                # a variant, expected to be an object
+  optional group typed_value {          # shredded fields for the variant object
+    required group event_type {         # shredded field for event_type
+      optional binary value;
+      optional binary typed_value (STRING);
+    }
+    required group event_ts {           # shredded field for event_ts
+      optional binary value;
+      optional int64 typed_value (TIMESTAMP(true, MICROS));
+    }
+  }
+}
+```
 
-# Reconstructing a Variant
+The group for each named field must be required.
 
-It is possible to recover a full Variant value using a recursive algorithm, where the initial call is to `ConstructVariant` with the top-level fields, which are assumed to be null if they are not present in the schema.
+A field's `value` and `typed_value` are set to null (missing) to indicate that the field does not exist in the variant.
+To encode a field that is present with a null value, the `value` must contain a Variant null: basic type 0 (primitive) and physical type 0 (null).
 
-```
-# Constructs a Variant from `variant_value`, `object`, `array` and `typed_value`.
-# Only one of object, array and typed_value may be non-null.
-def ConstructVariant(variant_value, object, array, typed_value):
-  if object is null and array is null and typed_value is null and variant_value is null: return VariantNull 
-  if object is not null:
-    return ConstructObject(variant_value, object)
-  elif array is not null:
-    return ConstructArray(array)
-  elif typed_value is not null:
-    return cast(typed_value as Variant)
-  else:
-    variant_value
-
-# Construct an object from an `object` group, and a (possibly null) Variant variant_value
-def ConstructObject(variant_value, object):
-  # If variant_value is present and is not an Object, then the result is ambiguous.
-  assert(variant_value is null or is_object(variant_value))
-  # Null fields in the object are missing from the reconstructed Variant.
-  nonnull_object_fields = object.fields.filter(field -> field is not null)
-  all_keys = Union(variant_value.keys, non_null_object_fields)
-  return VariantObject(all_keys.map { key ->
-    if key in object: (key, ConstructVariant(object[key].variant_value, object[key].object, object[key].array, object[key].typed_value))
-    else: (key, variant_value[key])
-  })
-
-def ConstructArray(array):
-  newVariantArray = VariantArray()
-  for i in range(array.size):
-    newVariantArray.append(ConstructVariant(array[i].variant_value, array[i].object, array[i].array, array[i].typed_value)
-```
+The series of objects below would be stored as:
```

Comment:
could you provide a link, (or maybe inline the values here) so it is easy t compare with the table?

Reply:
The values are the first column of the table. I rephrased to make it more clear.